### DISCUSSION TITLE
Native Image Committer Community Meeting 2021-10-20

### DISCUSSION THREAD

**Discussion by christianwimmer** [oracle/graal#3932] [Show and tell]
List of all past and upcoming meetings: #3933

## New and Noteworthy
* GraalVM 21.3 is released. Needed more last-minute bugfixes than we hoped.

* Please benchmark the new GC policy and provide feedback. Ideally with the latest daily snapshot build. Or with the 21.3 release and option `-H:InitialCollectionPolicy=Adaptive`.

* JDK 8 support is going to be deleted. Several PRs already switch some testing from JDK 8 to JDK 17 and do some cleanups:
  Switch SVM gates from JDK 8 to JDK 17: beda3cc66c6f6dac6e562d564ace4ac151c67379
  #3924
  
* Module system support:
  #3821
  #3748
  
* Robustness:
  #3874
  #3830
  
* Compatibility improvements:
  #3925
  #3854
  #3883
  #3792
  #3834

* JFR support:
  Add JFR support for threads: #3720
  #3743

* Image size reduction:
  #3928 (not ready for review or merging)
  
* Static analysis:
  #3926 (not ready for review or merging)
  #3801 (not ready for review or merging)
  #3666
  #3773
  
* Medium-term planning:
  Continuations/coroutines (support for Project Loom). Initial support was merged a year ago, but several bigger pieces and optimizations are still missing.
  
### Class initialization

* Default is initialization of classes at run time
* Options `--initialize-at-build-time` and `--initialize-at-run-time` work at the level of package names and class names: https://github.com/oracle/graal/blob/64389bc07d65594ef83fef7201eac76b9d94cd1f/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/classinitialization/ClassInitializationFeature.java#L83
* The core of the JDK is initialized at image build time, with a curated list of exceptions: https://github.com/oracle/graal/blob/1477a8266b38d61ee164fc3c79471bdf9dae8c78/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JDKInitializationFeature.java#L35
* A simple bytecode-level-analysis before the static analysis initializes some simple side-effect-free classes early: https://github.com/oracle/graal/blob/320fe549dbde377c4081481ef06b36c38db906ef/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/classinitialization/EarlyClassInitializerAnalysis.java#L114
* A inter-procedural analysis that uses static analysis results initializes some more complicated side-effect-free classes before compilation: https://github.com/oracle/graal/blob/153141d0986e49bb5cfd5fe23854e4d0635a8823/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/classinitialization/TypeInitializerGraph.java#L66

### Complete vs. incomplete classpath

* With the closed-world assumption, we can guarantee that no linking errors are thrown at run time. This is currently the default.
* `--allow-incomplete-class-path` turns the build-time check to a run-time check. Currently global for the whole image. Many libraries set the option.
* Several places use the flag to decide between "throw an exception in the image builder" or "synthesize Graal IR that throws an exception at run time". For example in the bytecode parser: https://github.com/oracle/graal/blob/cf56a336615fe9d6f403e26b7b250c79256e3297/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/SharedGraphBuilderPhase.java#L255

### How to improve?

* The default must be switched to "allow incomplete class path".
* Use modules as the granularity for configuration? Can we assume all libraries that need configuration are proper modules?
* Create a report all possible linking errors to all auditing?
* Some discussion in the comments of #2762
* #3929

## Possible deep dive topics for next meeting
Please send suggestions, or "upvote" a suggestion, by adding a comment to this discussion.
* _GC policy and GC performance_ (got a "vote" in the previous meeting)
* Reflection configuration: "query only" vs. "allow invocation" configuration for methods
* Conditional reflection configuration

**Comment by zakkak** [comment]
+1 for "GC policy and GC performance" as the deep dive topic of the next meeting

Thanks @christianwimmer !

**Comment by Sanne** [comment]
## Regarding "How to improve?" for incomplete classpath

To share here for those who didn't make it at the meeting: as I've pointed out during the call, Quarkus has no plans to require Java modules for its libraries, so we would highly prefer a flag which focuses on the package or jar file - or the higher granularity, to explicitly list all missing symbols which should not be treated as compile errors.

It was mentioned that other frameworks rely on "uber jars" which would make the jar-centric flag less useful; in fact Quarkus offers this option as well but it's not the preferred one, and we don't use this packaging format when building a native-image, and I hope that other frameworks would see the benefits of avoiding the uber jar approach as well, as its convenience is only a point for JVM based distributions of the application. I think we can all agree that when producing a native image it's an unnecessary step; we also highlighted the approach introduces severe maintainability concerns so while I don't deny it's nice to have such options, I don't think we should focus on this approach as being the one to recommend for production use cases.

Having a marker within the affected library (within its jar) seems like a sensible compromise: in case of the library being repackaged ("shaded") and combined with other code, at worst the marker would extend as well to the additional code it's combined with; in extreme cases it would propagate to the whole application, converging into the current semantics. But this would be a choice of granularity that framework developers and application packaging can control and opt-in (opt-out).

I also maintain that it would be great - for any specific library - to list which code is "allowed to be missing" with an exhaustive list. This would prevent such flags to unintentionally get applied on other cases it was not intended for, and helps with long term maintenance of such metadata correctness.

**Reply by christianwimmer** [reply]
PR with a proposed solution: https://github.com/oracle/graal/pull/4305
